llama : add pipeline parallelism support

slaren opened the pull request:
Pipeline parallelism improves batch processing performance when using multiple GPUs.

Changes:

- Graph inputs are allocated in the graph
  - Allocating inputs in the graph allows `ggml_backend_sched` to make multiple copies automatically, which is useful to reduce synchronization requirements with pipeline parallelism
  - Allocating inputs on the graph is more efficient since it allows their memory to be reused for other computations
  - Only the inputs actually required by the model are allocated
- Automatic batch splitting in `llama_decode`
  - `llama_decode` automatically splits the batches into multiple smaller batches if it is too big for the configured compute batch size
  - The largest batch size that can be submitted to `llama_decode` is still limited by `n_batch` to reduce the size of the logits and embeddings buffers
- Adds `n_ubatch` (`-ub` in the command line) to `llama_context_params` parameter
  - `n_batch` sets the size of the logits and embeddings buffer, which limits the maximum batch size passed to `llama_decode`
  - `n_ubatch` sets the maximum batch size for computation
  - By default `n_batch` is 4096, `n_ubatch` is 512
  - This allows current applications to take advantage of pipeline parallelism by setting a larger `n_batch` without having to update their logic
- Makes `llama_decode` asynchronous
  - Synchronization is done automatically on `llama_get_logits` and `llama_get_embeddings`
  - Adds `llama_synchronize` to force a synchronization manually, which can be useful when measuring the time of `llama_decode`
  - Applications can also take advantage of pipeline parallelism by making multiple calls to `llama_decode` without synchronizing
  - Note: `llama_timings` may not be accurate if the application does not synchronize immediately after calling `llama_decode`
- Uses a host buffer for the logits and embeddings (except when pooled) outputs, which improves the performance when copying the data from the GPU 
- Multi-threaded `ggml_get_rows` (still single-threaded when full offload)
- Adds an event interface to ggml-backend for synchronization between different backends
- CUDA only, other backends will need to implement the async and event interface to take advantage of pipeline parallelism
- Adds the build parameter `LLAMA_SCHED_MAX_COPIES`
  - This parameter configures the number of copies of the split inputs in `ggml_backend_sched` when using pipeline parallelism
  - Increasing this value allows more operations to be queued without requiring synchronization, and may improve performance on some systems at the expense of higher memory usage in the compute buffer

slaren committed:
llama : add pipeline parallelism support for batch processing with multiple CUDA GPUs

ggml-ci

phymbert commented:
Do you mind to also add `--ubatch-size` in server.cpp please ?

slaren committed:
server : add -ub, --ubatch-size parameter

slaren committed:
fix server embedding test

ggerganov commented:
Posting some results on 8x A100:

```bash
LLAMA_CUBLAS=1 make -j && CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 ./llama-bench -m models/codellama-7b/ggml-model-f16.gguf -m models/codellama-7b/ggml-model-q8_0.gguf -m models/codellama-7b/ggml-model-q4_0.gguf -ngl 99 -p 512,1024,2048,4096,8192 -b 8192
```

<details>

### old (sl/micro-batching) with ub=256:

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 8 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 2: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 3: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 4: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 5: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 6: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 7: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     | 12287.35 ± 404.42 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    | 18428.65 ± 238.85 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    | 24279.31 ± 225.94 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    | 27046.41 ± 54.81 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    | 24665.13 ± 61.06 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     71.59 ± 1.20 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   9222.69 ± 4.70 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  13907.00 ± 3.19 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    | 17866.35 ± 13.90 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    | 19697.05 ± 14.58 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  18544.07 ± 6.93 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     89.72 ± 2.34 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   8570.40 ± 4.00 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  12955.56 ± 9.16 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  16745.54 ± 4.98 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  18631.20 ± 2.73 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  17652.17 ± 2.92 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    128.17 ± 4.96 |

build: af789e7e (1861)


### new (x8 GPUs):

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 8 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 2: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 3: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 4: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 5: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 6: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 7: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |  8159.85 ± 12.87 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  12998.56 ± 7.68 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  14400.26 ± 4.17 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  15405.62 ± 6.74 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    | 12736.25 ± 14.53 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     72.82 ± 0.40 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6771.69 ± 3.19 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  10901.11 ± 3.95 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  12175.02 ± 6.43 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  13332.42 ± 6.95 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  11364.28 ± 9.12 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    112.28 ± 0.98 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6148.57 ± 2.28 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |   9968.76 ± 3.04 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  11159.57 ± 2.08 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  12344.41 ± 1.18 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  10671.60 ± 8.46 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    140.66 ± 1.65 |

build: 1ac668e4 (2410)


### new (x8 GPU, input on GPU):

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 8 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 2: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 3: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 4: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 5: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 6: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 7: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   8381.22 ± 2.50 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  13312.63 ± 5.10 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  14723.17 ± 5.76 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  15671.19 ± 2.66 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    | 12921.86 ± 12.28 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     72.84 ± 0.56 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6838.71 ± 3.35 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  11028.61 ± 3.34 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  12270.75 ± 2.90 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    | 13406.20 ± 56.99 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  11426.95 ± 8.87 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    112.35 ± 1.19 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6232.21 ± 2.04 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  10111.86 ± 1.85 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  11301.41 ± 2.67 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  12498.26 ± 3.58 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  10769.33 ± 9.18 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    141.03 ± 1.82 |

build: 1ac668e4 (2410)


### new (1x GPU):

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |  8668.24 ± 11.45 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  8251.25 ± 23.84 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  7592.36 ± 41.82 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  6574.66 ± 19.70 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |   5112.25 ± 4.68 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     74.77 ± 0.10 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6914.08 ± 2.59 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  6642.77 ± 33.08 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |   6198.12 ± 2.44 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |   5506.76 ± 5.54 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |   4439.04 ± 3.06 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    116.21 ± 0.25 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |  6297.40 ± 27.02 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  6050.55 ± 90.49 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  5675.65 ± 41.84 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  5087.17 ± 12.07 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |   4167.11 ± 1.26 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    147.54 ± 0.20 |

build: 1ac668e4 (2410)

### new (2x GPUs):

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 2 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |  8522.21 ± 22.01 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  10483.76 ± 7.50 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  11438.16 ± 2.57 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    | 10919.41 ± 10.02 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |   8927.34 ± 6.56 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     74.03 ± 0.26 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |  6945.64 ± 19.96 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  8627.59 ± 10.28 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  9494.31 ± 13.54 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |   9262.61 ± 9.53 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |   7829.32 ± 5.61 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    115.27 ± 0.42 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6311.05 ± 2.35 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |   7888.98 ± 5.88 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |   8723.62 ± 4.29 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |   8599.76 ± 3.14 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |   7368.73 ± 6.62 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    145.75 ± 0.83 |

build: 1ac668e4 (2410)

### new (4x GPU):

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 4 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 2: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 3: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |  8396.23 ± 11.47 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    | 12157.80 ± 22.92 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    | 14724.37 ± 15.19 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    | 15228.68 ± 11.29 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    | 12923.24 ± 17.47 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |     73.73 ± 0.32 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6905.29 ± 4.01 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |  10085.96 ± 2.53 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    | 12353.66 ± 16.41 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    | 13114.21 ± 10.77 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    | 11488.99 ± 11.45 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    114.56 ± 0.73 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 512     |   6262.35 ± 2.35 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 1024    |   9252.81 ± 5.95 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 2048    |  11325.01 ± 3.31 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 4096    |  12145.52 ± 7.69 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | pp 8192    |  10783.66 ± 9.79 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 | tg 128     |    144.35 ± 1.13 |

build: 1ac668e4 (2410)

### new (8x GPU), different ub:

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 8 CUDA devices:
  Device 0: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 2: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 3: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 4: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 5: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 6: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
  Device 7: NVIDIA A100 80GB PCIe, compute capability 8.0, VMM: yes
| model                          |       size |     params | backend    | ngl |    n_batch |   n_ubatch | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------: | ---------- | ---------------: |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 512     |  5232.13 ± 31.96 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 1024    |  5511.86 ± 42.66 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 2048    |   5473.01 ± 8.37 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 4096    |   5265.57 ± 0.49 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 8192    |   4705.85 ± 1.29 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | tg 128     |     72.80 ± 0.47 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 512     |  8636.13 ± 18.84 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 1024    |  9073.29 ± 47.20 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 2048    |  9367.69 ± 55.64 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 4096    |  8757.92 ± 10.17 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 8192    |   7627.35 ± 1.85 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | tg 128     |     73.29 ± 0.02 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 512     |  11214.35 ± 4.42 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 1024    |  13388.25 ± 5.76 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 2048    |  13341.21 ± 9.66 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 4096    |  12647.74 ± 3.82 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 8192    |  10552.74 ± 4.27 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | tg 128     |     72.85 ± 0.42 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 512     | 12821.46 ± 19.68 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 1024    | 14094.04 ± 11.03 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 2048    |  16229.25 ± 8.04 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 4096    |  14960.16 ± 8.95 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 8192    |  12659.27 ± 2.42 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | tg 128     |     73.24 ± 0.01 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 512     |   8219.68 ± 1.45 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 1024    |  13004.40 ± 6.54 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 2048    | 14414.14 ± 12.03 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 4096    |  15378.82 ± 5.37 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 8192    |  12703.17 ± 4.10 |
| llama 7B F16                   |  12.55 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | tg 128     |     73.25 ± 0.01 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 512     |   1590.86 ± 3.45 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 1024    |   1706.98 ± 3.98 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 2048    |   1715.72 ± 0.90 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 4096    |   1718.28 ± 0.15 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 8192    |   1655.76 ± 0.13 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | tg 128     |    112.64 ± 1.02 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 512     |   5176.40 ± 2.06 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 1024    |  5344.67 ± 17.03 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 2048    |  5635.26 ± 20.97 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 4096    |   5407.87 ± 5.04 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 8192    |   4981.06 ± 0.69 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | tg 128     |    112.61 ± 1.10 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 512     |   7123.31 ± 3.67 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 1024    |   8778.65 ± 6.01 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 2048    |  8824.47 ± 26.09 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 4096    |  8760.80 ± 21.81 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 8192    |   7715.87 ± 3.97 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | tg 128     |    112.63 ± 1.10 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 512     |   9386.99 ± 3.53 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 1024    |  10414.88 ± 2.07 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 2048    |  12309.25 ± 3.75 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 4096    | 11672.67 ± 18.97 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 8192    | 10410.35 ± 16.28 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | tg 128     |    112.66 ± 1.11 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 512     |   6766.39 ± 2.81 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 1024    |  10869.11 ± 3.03 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 2048    |  12131.52 ± 3.03 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 4096    | 13282.26 ± 10.68 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 8192    |  11323.44 ± 8.35 |
| llama 7B Q8_0                  |   6.67 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | tg 128     |    112.63 ± 1.15 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 512     |   2354.02 ± 7.16 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 1024    |   2521.67 ± 9.04 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 2048    |   2524.36 ± 2.02 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 4096    |   2506.08 ± 0.33 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | pp 8192    |   2372.50 ± 0.21 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         32 | tg 128     |    141.33 ± 1.70 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 512     |   4100.33 ± 1.43 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 1024    |  4249.72 ± 12.05 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 2048    |  4496.98 ± 13.69 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 4096    |   4364.25 ± 3.36 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | pp 8192    |   4102.67 ± 0.50 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |         64 | tg 128     |    141.26 ± 1.76 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 512     |   5756.20 ± 1.77 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 1024    |   7125.95 ± 6.30 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 2048    |  7221.97 ± 17.13 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 4096    |  7281.78 ± 15.77 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | pp 8192    |   6565.03 ± 3.42 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        128 | tg 128     |    141.39 ± 1.69 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 512     |   7985.23 ± 2.28 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 1024    |   8901.79 ± 1.78 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 2048    |  10627.38 ± 3.02 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 4096    | 10229.80 ± 14.34 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | pp 8192    |  9316.91 ± 12.32 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        256 | tg 128     |    141.22 ± 1.69 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 512     |   6139.22 ± 1.81 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 1024    |   9959.09 ± 2.58 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 2048    |  11129.99 ± 1.60 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 4096    |  12315.34 ± 0.82 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | pp 8192    |  10640.47 ± 7.05 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CUDA       |  99 |       8192 |        512 | tg 128     |    141.25 ± 1.72 |

build: 1ac668e4 (2410)

</details>

slaren commented:
You can usually get better performance with F16 models with `-ub 256`. I set the default to 512 because that's the current default and works better with quantized models. In the previous PR it was 256.

compilade reviewed (dismissed):
This breaks Mamba, but it's fixable.

To help with fixing, I managed to at least make `main` work with <https://github.com/compilade/llama.cpp/commit/3e06fcac0994dee80c9ac039846e671ad4699e38>, but `parallel` still triggers an assert with Mamba. I'll investigate further.

slaren commented:
@compilade thanks for testing, feel free to push your fixes here directly.

compilade committed:
llama : fix Mamba inference for pipeline parallelism

Tested to work correctly with both `main` and `parallel` examples.

slaren committed:
llama : limit max batch size to n_batch

slaren committed:
add LLAMA_SCHED_MAX_COPIES to configure the number of input copies for pipeline parallelism
default increase to 4 (from 2)

changing this value may improve performance for some systems, but increases memory usage

slaren committed:
fix hip build

slaren committed:
Merge remote-tracking branch 'origin/master' into sl/pipeline-parallelism

slaren committed:
fix sycl build (disable cpy_tensor_async)

slaren committed:
fix hip build

Engininja2 commented:
`hipLaunchHostFunc()` was added in [ROCm 5.4.1](https://github.com/ROCm/ROCm/blob/develop/CHANGELOG.md#rocm-541) as a beta API. Should that minimum version be added to the readme?

slaren commented:
`hipLaunchHostFunc` is not used at the moment. It may be useful in the future for adding support for pipeline parallelism between the CPU and CUDA backends, but at the moment it is just a stub.

compilade reviewed (commented):
I like that the input tensors are allocated in the graph, it's much cleaner than before.

But I think the new default size of the logits buffer might be too big.

slaren commented on llama.cpp:
```diff
@@ -12537,7 +12582,8 @@ struct llama_context_params llama_context_default_params() {
         /*.seed                        =*/ LLAMA_DEFAULT_SEED,
         /*.n_ctx                       =*/ 512,
-        /*.n_batch                     =*/ 512,
+        /*.n_batch                     =*/ 4096,
```
A bigger batch size is what allows pipeline parallelism to work. For example if the application submits a batch of 4096 tokens to `llama_decode`, these will be split in mini-batches of 256 tokens each (`n_ubatch`) and evaluated in as a pipeline in parallel between the available GPUs. It could be reduced to 2048 and still get most of the benefit.

compilade replied:
Pipeline parallelism by default seems desirable. But `n_batch` shouldn't exceed `n_ctx`. Even when passing `--ctx-size 128` to `main`, `n_batch` is still 4096 (from the default), and the `785.63 MiB` of logits are still allocated, even if they can't be used since a batch bigger than `n_ctx` will simply not be able to find a big enough KV slot (for Transformer-based models, at least).

slaren replied:
We can change `llama_new_context_with_model` to limit `n_batch` and `n_ubatch` to `n_ctx`, since there is no advantage to increasing it beyond that.

Ideally, we would set the defaults intelligently according to the hardware of the system, and only increase `n_batch` is the system can actually support pipeline parallelism, which requires several CUDA GPUs and full offload. However that would require deeper changes.

ggerganov replied:
> But `n_batch` shouldn't exceed `n_ctx`

This is not always the case. When using a model that does not utilize the KV cache (for example non-causal embedding model like BERT), we might want to run with `n_ctx = 0, n_batch = 8192`

With [4400153](https://github.com/ggerganov/llama.cpp/pull/6017/commits/4400153348693db3ef008dbf73c78a2a3ae6e39d) applied in such cases, we now have to allocate a KV cache due to `n_ctx = 8192` and it won't be used

Given that, should we revert the clamp change?

slaren replied:
I didn't realize that, I also added a check in `llama_new_context_with_model` that prevents `n_ctx` being zero. It would be good to clamp the value if possible to avoid wasting memory. Is there a simple way to check for this? Maybe `hparams.casual_attn`?

ggerganov replied:
`n_ctx = 0` actually would default to `n_ctx_train` which is OK.
By `n_ctx = 0` I meant to be some small value (e.g. `n_ctx = 16`) in order to not waste VRAM on the KV cache when it is not used

Yes, we can check `hparams.causal_attn`. When it is `false` then the graph should not use the KV cache.

Btw, note that we now have support for a special model GritLM that has `hparams.causal_attn == true`, but it supports at runtime to set `cparams.causal_attn == false` in order to generate embeddings. Currently, we will still reuse the same compute graph both for text generation and for embeddings, so the KV cache is necessary in both situations. Though, in some potential implementation we could build 2 different graphs - one with and one without KV cache.

But to keep things more simple, let's not consider that for now. Checking `hparams.causal_attn` should be good for now and when it is `false` it should imply that the KV cache will not be used by the compute graph of the model

compilade replied:
> Ideally, we would set the defaults intelligently according to the hardware of the system, and only increase n_batch if the system can actually support pipeline parallelism, which requires several CUDA GPUs and full offload.

Should the default `n_batch` be more conservative then, for now?
For models with large context (e.g when using `-c 16384`), where clamping doesn't reduce `n_batch`, a default `n_batch` of 4096 still uses around `500 MiB` more memory than the previous default of 512. For recurrent models, this is much bigger than the KV cache size.
At least for my use case this means I'll now have to specify `-b 512` to save memory, whereas users with many GPUs probably already set a bigger `n_batch` anyway (I might be wrong, though).

Ideally though, there should be a way to know how many logits per decode will actually be used at most, while still allowing bigger logical batches than that. This would solve the memory usage problem and thus a big default `n_batch` should be fine.
I think only perplexity calculations (in `perplexity` and `imatrix`) require the logits for the whole batch, while normally only as many logits as distinct sequences are required in the other examples. But `llama_get_logits_ith(ctx, i)` would no longer be equivalent to `llama_get_logits(ctx) + i*n_vocab`. I might explore this idea in a future PR.

slaren replied:
We could pre-allocate buffers for `n_max_seq` tokens only during initialization, and increase the size dynamically in `llama_decode` automatically if there is a request for more logits than that.

compilade replied:
> We could pre-allocate buffers for `n_max_seq` tokens only during initialization, and increase the size dynamically in `llama_decode` automatically if there is a request for more logits than that.

From what I understand, the logits aren't necessarily contiguous with each other in the output buffer, so yes pre-allocation and dynamic resizing could be done, but not until the way the layout of the logits is always made contiguous with no offset before the first used logits.

slaren replied:
We could definitely improve the handling of logits. Even `perplexity` and `imatrix` only need logits for `n_ctx/2`. We could also skip the computation of the output layer for the tokens where logits are not needed. IIRC there was a PR about this a while ago, but it was never merged.

slaren committed:
llama : limit n_batch and n_ubatch to n_ctx during context creation

slaren committed:
llama : fix norm backend

ggerganov committed:
batched-bench : sync after decode

ggerganov committed:
swiftui : sync after decode

ggerganov commented on llama.cpp:
```diff
@@ -8762,287 +8770,292 @@ static int llama_decode_internal(
+                case LLAMA_POOLING_TYPE_CLS:
+                case LLAMA_POOLING_TYPE_MEAN:
+                    {
+                        // FIXME: this may not work if the sequences are split into different batches
+                        GGML_ASSERT(n_tokens_all == n_tokens);
```
Yes, this does not work - probably need to add comment that `n_ubatch` should be big enough to fit the prompt when using non-causal masks

Repro:

```bash
./server -m ./models/bert-bge-small/ggml-model-f16.gguf --embedding -np 8 -ngl 999 -cb --port 6900 --log-format text -b 1024 -c 4096 -v -ub 512
```

```bash
# this hits the assert because the prompt is 513 tokens

python3 ./examples/server-embd.py
```

Fix in `server`:

```diff
diff --git a/examples/server/server.cpp b/examples/server/server.cpp
index 7a210b07..d98b0c71 100644
--- a/examples/server/server.cpp
+++ b/examples/server/server.cpp
@@ -1811,7 +1811,7 @@ struct server_context {
 
                         if (slot.embedding) {
                             // this prompt is too large to process - discard it
-                            if (slot.n_prompt_tokens > n_batch) {
+                            if (slot.n_prompt_tokens > params.n_ubatch) {
                                 slot.state = SLOT_STATE_PROCESSING;
                                 slot.command = SLOT_COMMAND_NONE;
                                 slot.release();
```

slaren replied:
This seems to be the only case where the distinction between `n_batch` and `n_ubatch` actually matters to the applications. Support for embedding models is relatively new, so hopefully it won't cause much confusion. Do we need to do anything other than fix the check in `server`?

ggerganov replied:
I think the only thing to do in `llama.cpp` is to move the assert at the beginning of `llama_decode_internal` and apply it if `!cparama.causal_attn`. Also add message to the assert: "non-causal attention requires n_ubatch >= n_tokens"

slaren replied:
Does the server also check that `n_ctx >= n_prompt_tokens`? `n_batch` and `n_ubatch` may be modified by `llama_new_context_with_model` if they are larger than `n_ctx`, so checking the `gpt_params` is not enough. We may need to add a `llama_n_ubatch` function to get the actual value.

slaren committed:
ggml : allow ggml_get_rows to use multiple threads if they are available

slaren committed:
check n_ubatch >= n_tokens with non-casual attention

slaren committed:
llama : do not limit n_batch to n_ctx with non-casual attn

ggerganov committed:
server : construct batch with size of llama_n_batch

ggerganov reviewed (approved):
I've updated the A100 numbers with the latest version

phymbert commented:
@ggerganov Would you please mind sharing master bench on x8 GPU ?

@slaren Am I reading wrong that all tg are actually slower ?

for example, on master 1gpu tg = 74.44tk/s for 7B f16 but only 71.87tk/s for 8gpu with these changes.

slaren commented:
`master` with 8x GPU would be a bit slower than master with 1x GPU, since there is no parallelism at all, but there is the additional overhead of copying data between GPUs. That's also the reason why tg is slower with multiple GPUs, single token generation cannot be parallelized.

slaren commented:
Are you looking at the results of `master` or `old`? `old` refers to the previous pipeline parallelism PR, which suffered from synchronization issues and didn't always produce correct results. Performance is better than master with multiple GPUs in every pp test. For example, 7B pp 2048 improved from 7113 t/s to 22931 t/s with 8 GPUs. Larger batch sizes improve performance further.

phymbert commented:
thanks for your reply 👍, I am looking for a comparison against master yes. We do not see the figures for x8 gpu. Is the tg increased also ? how much ?

as I am working on the performance CI pipeline, I want to understand what this PR brings. Thanks

slaren commented:
I noticed the thread sanitizer tests failing, but the errors don't make much sense to me. The errors are intermittent, running the jobs again eventually the test passes. I suspect that it is an issue with a particular runner.

slaren committed:
ggml_backend_cpu_graph_compute : fix return value when alloc fails

slaren committed:
llama : better n_batch and n_ubatch comment

slaren commented on llama.cpp:
```diff
@@ -6506,6 +6550,10 @@ struct llm_build_context {
                     ext_factor, attn_factor, beta_fast, beta_slow
                 );
                 cb(Kcur, "Kcur", il);
+
+                cur = llm_build_kv(ctx0, model, hparams, kv_self, gf,
+                        model.layers[il].wo, model.layers[il].bo,
+                        Kcur, Vcur, Qcur, KQ_mask, nullptr, n_ctx, n_tokens, kv_head, n_kv, 1.0f/sqrtf(float(n_embd_head)), cb, il);
```
I think this is something I messed up in a merge. Does this need to be here?

ggerganov replied:
No, it's not needed. Likely artifact from a merge, yes

slaren committed:
fix merge

slaren committed:
small fix

slaren committed:
reduce default n_batch to 2048

ggerganov commented:
@phymbert I've added multi-GPU results for `master` into the comment - hope this makes it more clear. With this PR we now have efficient pipeline parallelization which is important for increasing the throughput in multi-GPU systems

@slaren Yes, the thread sanitizer build failures are not related to our code

siavashmohammady66 commented:
Does it also support multiple GPU , when we are using Vulken SDK for compiling the llama.cpp code? (For the AMD GPU)
